### ISSUE TITLE
Allow nix store path other than `/nix` without hacky workarounds

### ISSUE THREAD

**Issue by Pandapip1** [NixOS/nix#10253]
**Is your feature request related to a problem? Please describe.**

I am trying to install `nix` on a SLURM HPC. The way the cluster is managed means that each lab is responsible for installing their own packages. Globally-installed packages are not allowed.

Currently, the recommended package manager is spack. Spack is like nix except significantly worse in every conceivable way. The only advantage it has is... being able to be installed at all.

I've asked the admins to globally install nix. They have NACKed my request. Installing in a chroot is not possible for various reasons relating to the way the cluster automatically mounts special filesystems and SLURM.

**Describe the solution you'd like**

Let me pick the nix store directory without ending up in a chicken-and-egg problem!

**Describe alternatives you've considered**

A single-user installation option that installs to `~/.bin` and has the store default to `~/.nix/store`. While I would prefer for labs to be able to use a common store, since I will be strongly encouraging flakes, this shouldn't be a big deal.

**Additional context**

The issue with chroots is that the normal filesystems (except for `/tmp`) are mounted readonly in SLURM jobs. Then, `/scratch`, a high-throughput low-redundancy filesystem is mounted. As far as I am aware. If I tried to use a chroot, I would end up with a readonly system that is incapable of outputting anything.

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by tgamblin** [comment]
> Spack is like nix except significantly worse in every conceivable way

wow.

**Comment by Pandapip1** [comment]
Okay. I was being a bit harsh to spack. But in it's current form, it offers no advantages over nix, and has some significant *dis*advantages.

Thoughts on the feature request itself?

**Comment by Ericson2314** [comment]
@Pandapip1 you know that the store dir is runtime configurable, right? You will have to give up `cache.nixos.org` but it should work fine. You can get a statically linked Nix (which is relocatable / can be dropped in any directory) to bootstrap rebuilding the world with your custom store dir.

See https://github.com/NixOS/nix/issues/9549 for a more ambitious long term solution

**Comment by Pandapip1** [comment]
> @Pandapip1 you know that the store dir is runtime configurable, right? You will have to give up cache.nixos.org but it should work fine. You can get a statically linked Nix (which is relocatable / can be dropped in any directory) to bootstrap rebuilding the world with your custom store dir.

No, I didn't. I searched the docs but couldn't find any way to install nix without either root or a chroot. Mind pointing me to a link? I'll check out the linked issue in the meantime.

**Comment by thufschmitt** [comment]
> No, I didn't. I searched the docs but couldn't find any way to install nix without either root or a chroot. Mind pointing me to a link? I'll check out the linked issue in the meantime.

Yes, this is quite hidden. All the info is in https://nixos.org/manual/nix/stable/store/types/local-store, but what you really care about here is the example at the top:

```console
$ nix build --store 'local?store=/tmp/my-nix/store&state=/tmp/my-nix/state&log=/tmp/my-nix/log' nixpkgs#hello
```

Or if you want something nicer than a command-line flag, add `store = local?store=/tmp/my-nix/store&state=/tmp/my-nix/state&log=/tmp/my-nix/log` to you Nix config file (`~/.config/nix/nix.conf`).

That can also be made the default at compile time (by passing `--with-store-dir=/somewhere/else` to the configure script).

Note however that like @Ericson2314 mentioned, changing the store directory means rebuilding the whole universe from scratch (in particular if each user has its own store).

**Comment by Pandapip1** [comment]
Doesn't this require you to have an active nix installation though? Again, I mentioned that this wasn't possible without having a chicken and egg problem, and this does seem like a chicken and egg problem.

**Comment by thufschmitt** [comment]
Ah, because your problem is bootstrapping! Sorry, I didn't get that.

The easiest way to bootstrap for you is probably to fetch a static build of Nix (you can get one from [here](https://hydra.nixos.org/job/nix/maintenance-2.21/buildStatic.x86_64-linux/latest-finished/download/2/nix)).

Or you can bootstrap it from source (but then you need to manually provide all the required dependencies, which is a bit more work).

**Comment by Pandapip1** [comment]
I've installed a static build of nix, but it really doesn't seem to be liking the configuration I've had to provide it. It is installed at `/ref/djslab/software/usr/local/bin/nix`, and I have the following custom configuration files:

```console
gavinjohn@login:~ $ cat .config/nix/nix.conf 
experimental-features = nix-command flakes
use-xdg-base-directories = true
require-sigs = false
store = /ref/djslab/software/nix-store
ssl-cert-file = /etc/ssl/certs/ca-bundle.crt
gavinjohn@login:~ $ cat .config/nix/channels 
https://nixos.org/channels/nixpkgs-unstable nixpkgs
```

But even a simple `nix run` command fails:

```console
gavinjohn@n001:~ $ nix run nixpkgs#hello
error: path '/ref/djslab/software/nix-store/nix/store/bv68i4l7g6s5sixxm0c9lcjdxhlkg0n5-source' is not in the Nix store
gavinjohn@n001:~ $ ll /ref/djslab/software/nix-store/nix/store/bv68i4l7g6s5sixxm0c9lcjdxhlkg0n5-source/
total 58
dr-xr-xr-x  8 gavinjohn djslab 157233804 Dec 31  1969 .
drwxr-sr-x  4 gavinjohn djslab 157247531 Mar 20 13:01 ..
-r--r--r--  1 gavinjohn djslab     36168 Dec 31  1969 CONTRIBUTING.md
-r--r--r--  1 gavinjohn djslab      1097 Dec 31  1969 COPYING
-r--r--r--  1 gavinjohn djslab       971 Dec 31  1969 default.nix
dr-xr-xr-x 15 gavinjohn djslab    989620 Dec 31  1969 doc
-r--r--r--  1 gavinjohn djslab      2916 Dec 31  1969 .editorconfig
-r--r--r--  1 gavinjohn djslab      4353 Dec 31  1969 flake.nix
-r--r--r--  1 gavinjohn djslab       662 Dec 31  1969 .gitattributes
-r--r--r--  1 gavinjohn djslab      3289 Dec 31  1969 .git-blame-ignore-revs
dr-xr-xr-x  4 gavinjohn djslab     63161 Dec 31  1969 .github
-r--r--r--  1 gavinjohn djslab       516 Dec 31  1969 .gitignore
dr-xr-xr-x  6 gavinjohn djslab    843729 Dec 31  1969 lib
-r--r--r--  1 gavinjohn djslab       950 Dec 31  1969 .mailmap
dr-xr-xr-x  3 gavinjohn djslab    799935 Dec 31  1969 maintainers
dr-xr-xr-x  7 gavinjohn djslab  13934186 Dec 31  1969 nixos
dr-xr-xr-x 20 gavinjohn djslab 107229126 Dec 31  1969 pkgs
-r--r--r--  1 gavinjohn djslab      5186 Dec 31  1969 README.md
lrwxrwxrwx  1 gavinjohn djslab        12 Dec 31  1969 .version -> lib/.version
gavinjohn@n001:~ $ nix run nixpkgs#hello
error: path '/ref/djslab/software/nix-store/nix/store/bv68i4l7g6s5sixxm0c9lcjdxhlkg0n5-source' is not in the Nix store
```

There's probably a config option I'm missing, but I really don't know what. Any pointers?

**Comment by Ericson2314** [comment]
@Pandapip1 make it

```
store = local?store=/ref/djslab/software/nix-store
```
otherwise it will try to do a chroot store.

**Comment by Pandapip1** [comment]
Now, whenever I try to do basically anything with nix, I get the following error:

```
error: path '/nix/store/<long hash>-flake-registry.json' is not in the nix store
```

This occurs whether I use the CLI option or have it in my config file. I've updated the paths, so the option is:

```
store = local?store=/ref/djslab/software/nix/store&state=/ref/djslab/software/nix/state&log=/scratch/djslab/nix/log
```

**Comment by thufschmitt** [comment]
> Now, whenever I try to do basically anything with nix, I get the following error:
> 
> ```
> error: path '/nix/store/<long hash>-flake-registry.json' is not in the nix store
> ```

I suspect you have a stale `flake-registry.json` file somewhere that is a symlink to `/nix/store/...` (leftover from the previous tries without the `store` option I guess). It would probably be under `~/.cache/nix/flake-registry.json`. Maybe try nuking `~/.cache/nix` altogether to make sure there isn't any other stray thing left

**Comment by thufschmitt** [comment]
(btw: I hope that the error above disappears if you remove the `flakes` experimental features)

**Comment by Pandapip1** [comment]
`rm -rf ~/.cache/nix` solved the issue! TYSM!

**Comment by Pandapip1** [comment]
Now I'm getting https://github.com/NixOS/nixpkgs/issues/299240
